core/src/nodes/raw_swarm.rs

```diff
@@ -1088,28 +1083,27 @@ where
         if event.would_replace() && has_dial_prio(&reach_attempts.local_peer_id, event.peer_id()) {
             if let Some(ConnectedPoint::Dialer { .. }) = reach_attempts.connected_points.get(event.peer_id()) {
                 if let ConnectedPoint::Listener { listen_addr, send_back_addr } = opened_endpoint {
-                    return (Default::default(), RawSwarmEvent::IncomingConnectionError {
+                    return RawSwarmEvent::IncomingConnectionError {
                         listen_addr,
                         send_back_addr,
                         error: IncomingError::DeniedLowerPriority,
-                    });
+                    };
                 }
             }
         }
 
         // Set the endpoint for this peer.
         let closed_endpoint = reach_attempts.connected_points.insert(event.peer_id().clone(), opened_endpoint.clone());
 
-        // Cancel any outgoing attempt to this peer.
-        let action = if let Some(attempt) = reach_attempts.out_reach_attempts.remove(&event.peer_id()) {
+        // We keep the current outgoing attempt because it may already have succeeded without us
+        // knowing. It is possible that the remote has already closed its ougoing attempt because
+        // it sees our outgoing attempt as a success.
+        // However we cancel any further multiaddress to attempt.
+        // TODO: cancel if we don't have dial priority? we should write tests first
```

Comment:
I think it is critical to do this as well; otherwise our dialing attempt will always replace the remote's connection, regardless of whether we have priority.